### PR TITLE
Add global navigation sidebar

### PR DESCRIPTION
**Pull request by kulyk**
This PR is a foundation of a new global navigation layout for Metabase. It basically adds the sidebar we used to have on the collection page to every Metabase page, so it becomes accessible everywhere (with a few exceptions).

The implementation is a bit rough (it's mostly just `CollectionSidebar`) and the visual appearance is not yet final and will be changed in the subsequent PRs.

### How to review

The branch used to be a prototype branch and we didn't follow a feature branch approach, so the first half of commits is done directly on the branch without their own PRs. It'd make sense to review them commit by commit. The other half is done in a feature branch manner where every commit is a squashed PR that was reviewed separately.

### To Verify
_A great way to test this is to run a fresh Metabase installation to ensure the initial flow looks good too_

1. Go through as many Metabase pages as possible (query builder, dashboards, collections, browse data, admin, etc.) and for each page:
2. Make sure you can open/collapse the navigation sidebar (by clicking the chevron icon next to the Metabase logo at the top left) and make sure the layout doesn't look broken and every feature is available
3. Make sure the top bar and the sidebar are hidden in two cases:
    * when editing a dashboard
    * when editing a model's query/metadata

**Known issues**
1. Active element highlighting: the sidebar uses `CollectionSidebar` and it highlights selected collections by ID. Because of that, if you open "Browse data" or archive any bookmark, the sidebar won't reflect that. This is going to be fixed by #21298
2. BulkActionsBar layout (the one that appears when you select a few collection items for a bulk archive/move)
3. Mobile layout

### Demo

CleanShot 2022-03-29 at 14 41 14@2x

### CODE REVIEW

File: frontend/src/metabase/redux/app.js [hunk 1]
```diff
@@ -28,6 +32,40 @@ const errorPage = handleActions(
...
+  return state.app.isNavbarOpen;
+}
+
+const isNavbarOpen = handleActions(
+  {
+    [OPEN_NAVBAR]: () => true,
+    [TOGGLE_NAVBAR]: isOpen => !isOpen,
+    [CLOSE_NAVBAR]: () => false,
+  },
+  checkIsSidebarInitiallyOpen(),
```
Comment (ranquild) [inline frontend/src/metabase/redux/app.js:65]: An idea for a follow-up PR - save the default value in a user setting.

File: frontend/src/metabase/App.jsx [hunk 2]
```diff
@@ -61,31 +83,128 @@ export default class App extends Component {
...
-          {errorPage ? getErrorComponent(errorPage) : children}
-          <UndoListing />
-          <StatusListing />
-        </div>
+        {errorPage ? (
+          getErrorComponent(errorPage)
+        ) : (
+          <>
+            {hasAppBar && (
+              <AppBar
```
Comment (ranquild) [inline frontend/src/metabase/App.jsx:182]: Can we move state from `App` component to redux and connect `AppBar` and another component to it? I believe that we should `App` is small is possible.

File: frontend/src/metabase/App.jsx [hunk 3]
```diff
@@ -61,31 +83,128 @@ export default class App extends Component {
...
+                isSidebarOpen={isSidebarOpen}
+                location={location}
+                onToggleSidebarClick={toggleNavbar}
```
Comment (ranquild) [inline frontend/src/metabase/App.jsx:185]: All of these props can be obtained by connect'ing

**Review by ranquild — APPROVED** [review]
I've tried different things and it seems to be working. I believe that `App.jsx` can be improved and simplified, but I'm not 100% sure about it. Great work!

**Comment by kulyk** [comment]
Thanks for your feedback @ranquild 🙌 If you don't mind, I'll proceed with `App.jsx` refactoring a bit later — I have a bunch of other PRs modifying it and it'll become painful to fix every single one of them. So I'm thinking about merging them first and then doing refactoring to make the `App.jsx` independent from redux

**Comment by ranquild** [comment]
> Thanks for your feedback @ranquild 🙌 If you don't mind, I'll proceed with `App.jsx` refactoring a bit later — I have a bunch of other PRs modifying it and it'll become painful to fix every single one of them. So I'm thinking about merging them first and then doing refactoring to make the `App.jsx` independent from redux

Sure

**Pull request merged**